client/src/components/PopoverOptionSet/PopoverOptionSet.js

```diff
@@ -0,0 +1,261 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { Button, Input, InputGroup, InputGroupAddon, Popover } from 'reactstrap';
+import classNames from 'classnames';
+import i18n from 'i18n';
+
+/**
+ * Fills a popover with buttons and a search box to filter the buttons.
+ * See the storybook for more details
+ */
+class PopoverOptionSet extends Component {
+  constructor(props) {
+    super(props);
+
+    this.handleToggle = this.handleToggle.bind(this);
+    this.handleSearchValueClear = this.handleSearchValueClear.bind(this);
+    this.handleSearchValueChange = this.handleSearchValueChange.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+
+    this.state = {
+      searchValue: ''
+    };
+  }
+
+  /**
+   * Pass toggle to parent (props requires a toggle function) and clear the search input
+   */
+  handleToggle() {
+    const { toggle } = this.props;
+
+    toggle();
+    this.handleSearchValueClear();
+  }
+
+  /**
+   * Handle click on clear button within search bar
+   */
+  handleSearchValueClear() {
+    this.setState(
+      { searchValue: '' }
+    );
+  }
+
+  /**
+   * Update the internal state on user input change
+   * @param {Event} event
+   */
+  handleSearchValueChange(event) {
+    this.setState(
+      { searchValue: event.target.value }
+    );
+  }
+
+  /**
+   * Handle key presses that are triggered while the popover is focused
+   *
+   * @param {SyntheticEvent} event
+   */
+  handleKeyDown(event) {
+    if (
+      // Ignore key presses in the search field...
+      event.target
+      && event.target.nodeName === 'INPUT'
+      // ...but allow "escape"
+      && event.key !== 'Escape'
+    ) {
+      return;
+    }
+
+    switch (event.key) {
+      case 'Escape':
+        this.handleToggle();
+        break;
+      default:
+    }
+  }
```

Comment:
Couldn't this be rewritten:

```js
handleKeyDown(event) {
  event.key === 'Escape' && this.handleToggle();
}
```

Reply:
Ah yep this method is a little overcomplicated - I was going to add a arrow key support but never got around to it. I'll update it.

Reply:
Yeah, we need to fix that. That's a really annoying rule.